Test/gsapTests/EventDrivenPrognoserTests.cpp

```diff
@@ -0,0 +1,212 @@
+ *
+ *   Tests for event-driven prognoser class.
+ *
+ *   @author    Chris Teubert
+ *   @version   1.2.0
+ *
+ *      Created: December 17, 2018
+ *
+ *   @copyright Copyright (c) 2018 United States Government as represented by
+ *     the Administrator of the National Aeronautics and Space Administration.
+ *     All Rights Reserved.
+ */
+
+#include <queue>
+#include <thread>
+
+#include "Datum.h"
+#include "EventDrivenPrognoser.h"
+#include "EventDrivenPrognoserTests.h"
+#include "Messages/ProgEventMessage.h"
+#include "Messages/ScalarMessage.h"
+#include "MockClasses.h"
+#include "ModelBasedEventDrivenPrognoserBuilder.h"
+#include "Models/PrognosticsModelFactory.h"
+#include "Observers/ObserverFactory.h"
+#include "Predictors/PredictorFactory.h"
+#include "Test.h"
+
+using namespace PCOE;
+using namespace PCOE::Test;
+
+const std::string SENSOR_SRC = "test_sensor_src";
+const std::string TRAJ_SRC = "test_traj_src";
+
+class TestComm : public IMessageProcessor {
+public:
+    TestComm(MessageBus& bus) : bus(bus) {
+        bus.subscribe(this, SENSOR_SRC);
+    }
+
+    void init(std::map<MessageId, Datum<double>> data) {
+        Require(data.size() > 0, "Must include atleast one data point to publish");
+        using std::chrono::milliseconds;
+        time_point time = time_point::min();
+
+        for (auto&& datapt : data) {
+            double dataValue = datapt.second.get();
+            time_point timept = time_point(milliseconds(datapt.second.getTime()));
+            time = timept > time ? timept : time;
+            ScalarMessage<double>* message =
+                new ScalarMessage<double>(datapt.first, SENSOR_SRC, timept, dataValue);
+            bus.publish(std::shared_ptr<Message>(message));
+        }
+    }
+
+    ProgEvent publish(std::map<MessageId, Datum<double>> data) {
+        Require(data.size() > 0, "Must include atleast one data point to publish");
+        using std::chrono::milliseconds;
+        time_point time = time_point::min();
+
+        // Setup Messages
+        std::vector<ScalarMessage<double>*> messages;
+        for (auto&& datapt : data) {
+            double dataValue = datapt.second.get();
+            time_point timept = time_point(milliseconds(datapt.second.getTime()));
+            time = timept > time ? timept : time;
+            messages.push_back(
+                new ScalarMessage<double>(datapt.first, SENSOR_SRC, timept, dataValue));
+        }
+
+        // Setup request
+        Request request(time);
+        unique_lock requests_lock(processMessageMut);
+        requests.push(&request);
+        // Publish Message
+        for (auto message : messages) {
+            bus.publish(std::shared_ptr<Message>(message));
+        }
+        requests_lock.unlock();
+
+        unique_lock lock(request.mut);
+        request.waiting = true;
+        while (request.waiting) {
+            request.condition.wait(lock);
+        }
+
+        // Process Result
```

Comment:
Re-lock `requests_lock` here to make sure we are accessing `currentResult` safely.